plugins/inputs/t128_tank/reader.go

```diff
@@ -172,10 +224,58 @@ func (r *Reader) read(mainCtx context.Context, readCtx context.Context, starting
 	r.readDone <- err
 }
 
-func (i *IndexedMessage) WithTimestamp(timestamp int) []byte {
-	parts := strings.Fields(string(i.Message))
-	parts[len(parts)-1] = strconv.Itoa(timestamp)
-	return []byte(strings.Join(parts, " "))
+func (r *Reader) readFromTank(readCtx context.Context, startingIndex index) (err error) {
+	cmdArgs := []string{
+		"/opt/128technology/bin/tank-cli",
+		"-b",
+		fmt.Sprintf("%s:%d", r.tankAddress, r.tankPort),
+		"-t",
+		r.topic,
+		"consume",
+		"-F",
+		"seqnum,content",
+		startingIndex.string(),
+	}
+
+	cmd := r.tankReadCmdCtx(readCtx, cmdArgs[0], cmdArgs[1:]...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("unable to pipe %s tank output: %s", r.topic, err)
+	}
+
+	tankReader := bufio.NewReader(stdout)
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("unable to start %s tank stream: %w", r.topic, err)
+	}
+
+	var parseErr error
+	var messages []*IndexedMessage
+
+	for {
+		select {
+		case <-readCtx.Done():
+			r.log.Errorf("%s read routine done", r.topic)
+			return nil
+		default:
+		}
+
+		messages, parseErr = parseLines(tankReader, r.topic)
+		if parseErr != nil {
+			r.log.Errorf("encountered error while parsing lines in %s output, will not attempt to parse more lines: %v", r.topic, parseErr)
+			return parseErr
+		}
+		var collectedMessages []IndexedMessage
+		for _, message := range messages {
+			if message.IsValid() {
+				collectedMessages = append(collectedMessages, *message)
```

Comment:
Not great that we're copying all these messages here. Can we drop invalid messages inside `parseLines` and just pass everything back to `sendChan` here?